### PR TITLE
Improve XMPP action, add support for java-pinning and fix #967

### LINKED ISSUE #967: XMPP 100% CPU after lost connection to server
I have XMPP settings and it working OK for the first run. But after openhab xmpp client lost connection to server and trying to reconnect i have 100% CPU and not working XMMP client.

in logs tons:

```
7:56:32.134 ERROR o.o.a.x.internal.XMPPConnect[:207]- XMPP re-connection failed, giving up: null
07:56:47.181 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:56:47.184 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
07:58:00.643 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:58:03.561 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
07:58:12.378 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:58:12.380 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
07:58:17.719 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:58:17.724 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
07:58:57.629 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:19.389 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:19.400 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:19.429 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:19.443 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:19.459 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:19.467 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:19.599 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:22.644 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:22.656 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:22.658 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:22.847 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:25.812 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:32.387 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
07:59:32.460 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:00:07.486 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:00:07.489 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
08:00:35.391 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:00:35.393 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
08:00:38.274 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:00:38.276 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
08:00:38.422 ERROR o.o.a.x.internal.XMPPConnect[:121]- Could not establish connection to XMPP server 'jabbim.com:5222': No response from the server.
08:00:38.423 ERROR o.o.a.x.internal.XMPPConnect[:207]- XMPP re-connection failed, giving up: null
08:00:41.367 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:00:41.369 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
08:01:15.883 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:02:44.828 ERROR o.o.a.x.internal.XMPPConnect[:121]- Could not establish connection to XMPP server 'jabbim.com:5222': No response from the server.
08:02:44.831 ERROR o.o.a.x.internal.XMPPConnect[:207]- XMPP re-connection failed, giving up: null
08:02:47.878 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:02:47.880 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
08:04:35.806 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:04:35.808 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
08:04:53.895 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
08:04:53.897 INFO  o.o.a.x.internal.XMPPConnect[:205]- XMPP re-connection succeeded.
08:04:53.900 INFO  o.o.a.x.internal.XMPPConnect[:199]- XMPP connection has been closed on error: Remote host closed connection during handshake
```

Server: jabbim.com:5222

### PR DESCRIPTION
**Pull request by Flowdalic**
Now that the XMPP action already supports a setting for TLS, it's time to add TLS pinning support. The idea is simply that most users won't be able to run openHAB with a default SSLContext that authenticates their certificate (leave alone a hostname checker). Instead we use a simple, lightweight, open source library, to add support for certificate (or pubkey) PINs, that can easily be created and configured. For more information refer to the corresponding new openhab config section at https://github.com/Flowdalic/openhab/blob/60cb377b8d10f37de8f90d22a18b8f03b91c14ab/distribution/openhabhome/configurations/openhab_default.cfg#L104

I also expect f94a2555e3778b084a20c9fea5268ac301be8d56 to fix #967.

Note that  d3a675a raises the source/target level to 1.7. After I got now answers to my question posted on the openhab mailing list about that topic [1], which I find a bit disappointing. I decided to do what I deemed necessary. Side note: The situation regarding the communication between the openhab maintainers and contributors could be improved. I'd like to see more collaboration and communication, ideally using IRC.

1: https://groups.google.com/forum/#!searchin/openhab/target$20level/openhab/fmEk_fKGiyA/naWTM7t4kzcJ

### CODE REVIEW

**Comment by Flowdalic** [comment]
That build failure doesn't look related to this PR

```text
channel stopped
Sending e-mails to: teichsta@gmail.com null
ERROR: 250 OK id=1YBBfO-0001Kb-8J

com.sun.mail.smtp.SMTPSendFailedException: 250 OK id=1YBBfO-0001Kb-8J
;
  nested exception is:
	com.sun.mail.smtp.SMTPAddressFailedException: 501 <null>: recipient address must contain a domain

	at com.sun.mail.smtp.SMTPTransport.sendMessage(SMTPTransport.java:1112)
	at javax.mail.Transport.send0(Transport.java:195)
	at javax.mail.Transport.send(Transport.java:124)
	at hudson.tasks.MailSender.execute(MailSender.java:116)
	at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.cleanUp(MavenModuleSetBuild.java:1038)
	at hudson.model.Run.execute(Run.java:1710)
	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:509)
	at hudson.model.ResourceController.execute(ResourceController.java:88)
	at hudson.model.Executor.run(Executor.java:246)
Caused by: com.sun.mail.smtp.SMTPAddressFailedException: 501 <null>: recipient address must contain a domain

	at com.sun.mail.smtp.SMTPTransport.rcptTo(SMTPTransport.java:1686)
	at com.sun.mail.smtp.SMTPTransport.sendMessage(SMTPTransport.java:1098)
	... 8 more
Putting comment on the pull request
Finished: FAILURE
```

**Comment by teichsta** [comment]
Hi @Flowdalic,

> The situation regarding the communication between the openhab maintainers and contributors could be improved. I'd like to see more collaboration and communication

I am sad to hear that you are disappointed. I try hard to follow as much conversations, issues (in at least two Forums), PRs, Social-Media Interactions as possible but sometimes it's not enough because the project grows and grows (which i am very happy with). And please don't forget openHAB is our spare time project … (luckily) we do have day jobs :-)

>  ideally using IRC.

I understand your point but I do really prefer asynchronous communication over the synchronous. Since openHAB is worldwide spreaded project at any time there is somebody awake and actively asking questions and he expects the same responsiveness … 

I hope for your understanding,

Thomas E.-E.

**Comment by teichsta** [comment]
> Note that d3a675a raises the source/target level to 1.7. After I got now answers to my question posted on the openhab mailing list about that topic [1],

I've answered your question on the group. Hope that helps.

Best, Thomas E.-E.

**Comment by Flowdalic** [comment]
Pushed new version that does only set 1.7 for the XMPP action.

**Comment by teichsta** [comment]
Thanks, which of your changes requires Java 1.7? Best, Thomas E.-E.

**Comment by Flowdalic** [comment]
Currently none, but Smack requires Java7 and I'd like to use try/multi-catch in situations like https://github.com/openhab/openhab/blob/master/bundles/action/org.openhab.action.xmpp/src/main/java/org/openhab/action/xmpp/internal/XMPPConnect.java#L161

**Comment by teichsta** [comment]
ok, but than we would have to increase the execution environment given in MANIFEST.MF in order to ensure this …

**Comment by Flowdalic** [comment]
@teichsta pushed 7753ce96

**Comment by teichsta** [comment]
besides 94a255 are there any other commits to cherry-pick into 1.6.2?

**Comment by Flowdalic** [comment]
> besides 94a255 are there any other commits to cherry-pick into 1.6.2?

f94a255 is the only really important bugfix. The others could go into the 1.6 branch too, but there is no urgent need to do so I guess.

**Comment by teichsta** [comment]
ok, thanks!
